scripts/pre_commit_linter.py

```diff
@@ -49,16 +49,17 @@
 
 Note that the root folder MUST be named 'oppia'.
  """
```

Comment:
I'm realizing that this file will be really hard to test. Do you have a strategy set up for which behaviors you plan to ensure works before/after the migration? Likewise for the other bash scripts we will need to convert?

Reply:
I have shared the details on testing the scripts on the follow up doc. But this file is indeed really hard to test. I have been thinking the whole point of making our codebase 100% tested in the first place was to be sure that in our test suite if any failures that appear after having tools rewrite the code(for Python 2 -> 3 conversion) are actual bugs in the tools and not in our code. We will achieve this at the end of the second milestone. Hence if it is at all possible the testing(especially the hard ones) can be done in post-gsoc period.

Reply:
Unfortunately not. Improving backend coverage to 100% is the primary goal of this project. Once that is done, we will be able to migrate to python 3 with more confidence.

I suggest instead outlining the "behaviors" that you want to test, rather than every line that should be tested?

Reply:
OK so I guess we can run the scripts with different flags at each time and ensure that they work properly? This will ensure that the scripts run properly against the files in the codebase. The behaviour includes all the functions that the script was supposed to do. This can be a manual process and can be noted in a doc or spreadsheet.

Reply:
No, we'll still need automated tests to make sure things are working correctly. Part of testing code is making the code testable in the first place, and this file is admittedly rather far from that. I have a few ideas of how to make it easier to test; for example, I'd like to write classes that hold onto the code/files/etc. for each language we test. We would then be able to test those classes separately and the business logic can focus on actually linting (currently, it is very mixed).

In short: there needs to be work to refactor this file into an easier-to-test structure if we want to achieve 100% backend coverage (and making code actually *testable* is part of that process).

I'd be happy to help you come up with ideas, I actually have a local PR for the "per-language data" in works locally.